lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,255 @@
+class DoConcurrentLabelEnforce {
+public:
+  DoConcurrentLabelEnforce(
+      parser::Messages &messages, std::set<parser::Label> &&labels)
+    : messages_{messages}, labels_{labels} {}
+  template<typename T> bool Pre(const T &) { return true; }
+  template<typename T> bool Pre(const parser::Statement<T> &statement) {
+    currentStatementSourcePosition_ = statement.source;
+    return true;
+  }
+  template<typename T> void Post(const T &) {}
+
+  // C1138: branch from within a DO CONCURRENT shall not target outside loop
+  void Post(const parser::GotoStmt &gotoStmt) { checkLabelUse(gotoStmt.v); }
+  void Post(const parser::ComputedGotoStmt &computedGotoStmt) {
+    for (auto &i : std::get<std::list<parser::Label>>(computedGotoStmt.t)) {
+      checkLabelUse(i);
+    }
+  }
+  void Post(const parser::ArithmeticIfStmt &arithmeticIfStmt) {
+    checkLabelUse(std::get<1>(arithmeticIfStmt.t));
+    checkLabelUse(std::get<2>(arithmeticIfStmt.t));
+    checkLabelUse(std::get<3>(arithmeticIfStmt.t));
+  }
+  void Post(const parser::AssignStmt &assignStmt) {
+    checkLabelUse(std::get<parser::Label>(assignStmt.t));
+  }
+  void Post(const parser::AssignedGotoStmt &assignedGotoStmt) {
+    for (auto &i : std::get<std::list<parser::Label>>(assignedGotoStmt.t)) {
+      checkLabelUse(i);
+    }
+  }
+  void Post(const parser::AltReturnSpec &altReturnSpec) {
+    checkLabelUse(altReturnSpec.v);
+  }
+  void Post(const parser::ErrLabel &errLabel) { checkLabelUse(errLabel.v); }
+  void Post(const parser::EndLabel &endLabel) { checkLabelUse(endLabel.v); }
+  void Post(const parser::EorLabel &eorLabel) { checkLabelUse(eorLabel.v); }
+  void checkLabelUse(const parser::Label &labelUsed) {
+    if (labels_.find(labelUsed) == labels_.end()) {
+      messages_.Say(currentStatementSourcePosition_,
+          "control flow escapes from DO CONCURRENT"_err_en_US);
+    }
+  }
+
+private:
+  parser::Messages &messages_;
+  std::set<parser::Label> labels_;
+  parser::CharBlock currentStatementSourcePosition_{nullptr};
+};
+
+// Find a canonical DO CONCURRENT and enforce semantics checks on its body
+class FindDoConcurrentLoops {
+public:
+  FindDoConcurrentLoops(parser::Messages &messages) : messages_{messages} {}
+  template<typename T> constexpr bool Pre(const T &) { return true; }
+  template<typename T> constexpr void Post(const T &) {}
+  template<typename T> constexpr bool Pre(const parser::Statement<T> &) {
+    return false;
+  }
+  bool Pre(const parser::DoConstruct &doConstruct) {
+    if (std::get<std::optional<parser::LoopControl>>(
+            std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t)
+                .statement.t)
+            .has_value() &&
+        std::holds_alternative<parser::LoopControl::Concurrent>(
+            std::get<std::optional<parser::LoopControl>>(
+                std::get<parser::Statement<parser::NonLabelDoStmt>>(
+                    doConstruct.t)
+                    .statement.t)
+                ->u)) {
```

Comment:
This would be easier to read if `std::get<std::optional<parser::LoopControl>>(std::get<parser::Statement<parser::NonLabelDoStmt>>(doConstruct.t).statement.t)` were extracted into a variable.

Reply:
I'm not in favor of introducing any variable names since it seems to always lead to review drag.

Reply:
I don't know what "review drag" is, but that expression really would be more readable if the reference to the extracted `std::optional` were made a local auto reference in an `if` predicate expression and its body, yes?  Otherwise, there's a multi-line common subexpression that leaves the reader wondering why it appears twice and whether the two instances are subtly distinct.

```
if (const auto &loopControl{std::get<std::optional<parser::LoopControl>>(...)}) {
  // use loopControl->
}
```